docs/design/unified_sidecar_and_process_monitor.md

```diff
@@ -0,0 +1,118 @@
+# Design on Combining the Main Container and the Sidecar Container on Kubernetes
+
+## Metadata
+
+* Authors: @brownleej
+* Created: 2020-12-21
+* Updated: 2020-12-21
+
+## Background
+
+We currently make use of three different containers in our Kubernetes deployment: a main container that runs fdbmonitor and fdbserver, an init container that generates the monitor conf and cluster file for the main container, and a sidecar container that provides updates to the monitor conf and cluster file for the main container. Running additional containers creates resource overhead and requires more complex interactions between containers. We would like to find a way to combine the function in the two containers for a simpler set of interactions.
+
+## General Design Goals
+
+* Remove the init container from our deployments
+* Remove any use of the sidecar container outside of copying new binaries during upgrades
+
+## Proposed Design
+
+We will build a new process called `fdb-kubernetes-monitor` that is designed to run fdbserver processes itself. It will take a configuration file that specifies the processes to launch, including explicit environment variable substitutions, and will not have any mechanism for fdbmonitor conf file management. It will also have APIs to copy files into another folder, and to check the hashes of files, like the sidecar does. This process will not be a drop-in replacement for either the sidecar or fdbmonitor, and will be designed specifically for use within the Kubernetes operator. It will be available to use for other purposes, such as injecting FDB client libraries into a client application.
+
+fdb-kubernetes-monitor will watch the configuration file for changes and automatically reload the desired process state, but will not automatically kill processes. It will watch for its child processes to die and relaunch them with the latest configuration. It will also do appropriate signal handling for SIGKILL and SIGTERM images in accordance with Docker best practices. The configuration file will be designed such that all of the processes of a single machine class can share a configuration file, with environment variables for command-line arguments that vary between processes. fdb-kubernetes-monitor will have an option for how many fdbserver processes to spawn, and the configuration file will support defining arguments that vary between different processes in the pod based on the process number.
```

Comment:
Could you add some more information how a change is propagated? The current design states `..., but will not automatically kill processes. It will watch for its child processes to die and relaunch them with the latest configuration.` Why should the child process die?

Reply:
Also, why watch for changes to configuration file instead of only re-reading it whenever it's time to take an action?  (Since it won't be automatically killing processes)

Reply:
That's a good point. I think we're better off not reloading the configuration automatically, so instead we would want an API to poll for (but not reload) changes, and an API to reload the changes immediately before a kill. This would also help limit the risk of an organic process failure causing it to come up on config that is different from the rest of the processes.

Reply:
I would prefer a solution where we don't need an API and can rely on the mechanisms from Kubernetes e.g. update the status of the object with the current version(s), config-file versions etc. so that we don't need to use an API. That should make the system more reliable if there is a network partition between the operator and some Pods.

Reply:
We're going to need connectivity between the operator and the pods in order to do the kills, and I don't think there's anything in the Kubernetes API that could give us kills that are as reliable and synchronized as doing the kills through the CLI or client library.

Reply:
That's right but I see them as different things. Pushing/checking the "version" of the binaries and ConfigMap requires the operator to talk to all Pods (e.g. needs connectivity to all of them) whereas the `fdbcli --exec `kill ..` command only requires access to one (?) or a few Pods.

Reply:
I believe the `kill` command requires the operator to be able to reach every pod that it is trying to kill.

Reply:
I think you're right about the core idea of reducing this interaction and relying on the Kubernetes API as a means of expressing this state. I'll work on modifying this design to accommodate that.

Reply:
I updated this to use the Kubernetes API to coordinate this state. As part of that, the design still has the launcher proactively watch for changes, since I don't think that a flow where the operator tells the launcher to reload configuration will work as well without a direct API to the launcher.